### PR TITLE
Refactor makefiles

### PR DESCRIPTION
**Pull request by mkristien**
## Description
Reunite makefiles and make them more modular with clearer configuration structure.

[ *Click up link*](https://app.clickup.com/t/2mnvcz)

## Changes
- move partial makefile scripts into `utils/build/`
- unite the main and the testing makefiles
- testing targets and variables have test annotation, e.g. `T_CFLAGS`
- [wiki page](https://github.com/Hyp-ed/hyped-2020/wiki/Makefiles) with Makefile documentation
- Partial support for CCLS (a language indexing engine better than VS Code's intellisense). There is now a [task](https://app.clickup.com/t/44fxtj) to introduce CCLS properly, including writing a wiki guide.

### CODE REVIEW

File: utils/build/libs.mk
Comment (ghost) [inline utils/build/libs.mk]: Typo in "RAPIDJSOM"
```suggestion
# RAPIDJSON is a header only lib, no compilation needed
```

File: utils/build/test.mk
Comment (neilweidinger) [inline utils/build/test.mk]: How come `NOLINT` get moved out to `test.mk`? Isn't overriding the linter applicable to everyone and not just testing?

File: utils/build/libs.mk
Comment (ghost) [inline utils/build/libs.mk]: Should we replace this incase we ever change the location of the lib file?
```suggestion
	$(Verb) tar -zxvf $@.tar.gz -C $(LIBS_DIR) > /dev/null
```

File: utils/build/libs.mk
Comment (ghost) [inline utils/build/libs.mk]: Same with the other, would it be better to replace this with a variable?
```suggestion
	$(Verb) tar -zxvf $@.tar.gz -C $(LIBS_DIR) > /dev/null
```

File: Makefile
Comment (neilweidinger) [inline Makefile]: What is the point of this target, and what was the reason for the removal of `cleanlint` and `cleantest`?

File: hyped.code-workspace
Comment (neilweidinger) [inline hyped.code-workspace:13]: Is it possible to maybe have this file moved somewhere out of the root directory and maybe into `utils/` or something?

File: utils/build/test.mk
Comment (mkristien) [reply utils/build/test.mk]: This should be removed, it was mistakenly left here after copy-paste from the original testing makefile.

File: utils/build/build.mk
Comment (mkristien) [inline utils/build/build.mk:3]: What is the justification for these changes?

File: utils/build/build.mk
Comment (mkristien) [inline utils/build/build.mk]: This has to be `=` not `:=`, as the `DEPFLAGS` are used inside a recipe where those `$*` will be resolved "dynamically.

Comment (mkristien) [reply utils/build/build.mk]: I have reverted this back to `=`,, the main thing did not even build for me with `:=`

File: utils/build/build.mk
Comment (mkristien) [reply utils/build/build.mk:3]: Also, if this is really desired, you can use `+=` instead.

File: hyped.code-workspace
Comment (brano2) [reply hyped.code-workspace:13]: We [discussed this before](https://github.com/Hyp-ed/hyped-2020/pull/10#issuecomment-555437995) :smile: 
(Also, I think it would belong under a separate PR if we were to move it.)

File: utils/build/test.mk
Comment (ghost) [inline utils/build/test.mk]: This currently regenerates testrunner in root directory. 

```suggestion
T_TARGET  := $(TEST_DIR)/testrunner
```

File: hyped.code-workspace
Comment (neilweidinger) [reply hyped.code-workspace:13]: Yea I remember that, but it would be nice if we could tidy up our root directory a bit. If it's just a matter of updating the relative paths it doesn't sound like too much too change imo.

File: utils/build/build.mk
Comment (neilweidinger) [reply utils/build/build.mk:3]: The problem was that since the main `Makefile` includes both `config.mk` and `build.mk`, and the flag to use c++11 is added in `config.mk` and `config.mk` gets included before `build.mk`, the `:=` operator in `build.mk` would simply overwrite the c++11 flag and would cause the compilation to fail.

File: hyped.code-workspace
Comment (brano2) [reply hyped.code-workspace:13]: Not sure if it's just that. We'd need to test it. Still, it's not a good practice to just tack unrelated stuff onto a PR. Better to have a tiny separate one. Do you want to add a task for it on ClickUp?

File: Makefile
Comment (brano2) [inline Makefile:7]: @mkristien did I break it with the comments after the variables? 😅

File: Makefile
Comment (brano2) [inline Makefile:74]: @mkristien what is the CCLS stuff about again? Maybe add it to the changelist in the PR's description?

File: Makefile
Comment (brano2) [inline Makefile]:
```suggestion
HELPERS_DIR  := utils/build
include $(HELPERS_DIR)/config.mk
include $(HELPERS_DIR)/build.mk
include $(HELPERS_DIR)/libs.mk
include $(HELPERS_DIR)/test.mk
```
Not v important but I read this section 3 times before fully understanding it. (At a first glance, I knew it was importing the other makefiles. I thought the `HELPER` variable was unrelated. On second reading, I thought you used `HELPER` just to avoid typing `utils/build` many times. Then I noticed that you actually call the other makefiles helpers.)

File: Makefile
Comment (brano2) [inline Makefile]:
```suggestion
# Default build configuration
```

File: Makefile
Comment (brano2) [inline Makefile]:
```suggestion
# Include helper files
```

File: Makefile
Comment (brano2) [inline Makefile]: Shouldn't this be `.PHONY: help`? If not, what is `.PHONE:`?

File: utils/build/config.mk
Comment (brano2) [inline utils/build/config.mk]:
```suggestion
# This makefile contains common configuration that should be use throughout the project.
# It defines ROOT variable which points to the root of this project.
```

File: utils/build/libs.mk
Comment (brano2) [inline utils/build/libs.mk]:
```suggestion

# Libaries for generating TARGET
```

File: utils/build/test.mk
Comment (brano2) [inline utils/build/test.mk]:
```suggestion
# Defines T_SRCS
```

File: utils/build/test.mk
Comment (brano2) [inline utils/build/test.mk]:
```suggestion
# Dependency checks
```

File: utils/build/test.mk
Comment (brano2) [inline utils/build/test.mk]:
```suggestion
# Pass this option to run static checker with specified severity.
# E.g. make static STATIC_ENABLE=style
```

**Review by brano2 — COMMENTED** [review]
I've added a few minor suggestions, but they are mostly minor style things. They improve readability for me personally, but if you find it worse, feel free to reject. I still need to review the helper makefiles. Will do that separately. Sorry for the delay.

File: Makefile
Comment (mkristien) [reply Makefile:7]: Yes. So it was you...

With the same line comments `VERBOSE` is not `1` but `1<several spaces here>` which breaks the `if` statements later on.

**Comment by mkristien** [comment]
Is there a way to accept multiple suggestions in a single commit?

File: Makefile
Comment (mkristien) [reply Makefile:74]: CCLS is a language indexing engine, think c++ intellisense. This one is, however, much better. This target tells the engine how things are compiled, where are include folders, which macro variables are defined,... Then, VSCode can navigate the source code much better. Tracing symbol references, go-to-definitions, highlighting code based on `#ifdef LINUX`,...

It requires some setup/installation. Everybody can get it done using google, but a guide might be more useful. However, this PR is dragging for ages now and I am not up to writing the guide soon. Feel free to remove it if you do not want to create clutter in the Makefile.

File: utils/build/test.mk
Comment (mkristien) [reply utils/build/test.mk]: That was also the original way of running the test. Let's hope moving this won't break the scripts running the testrunner.

**Comment by brano2** [comment]
> Is there a way to accept multiple suggestions in a single commit?

Yes. There are two buttons on each suggestion box. Just click the one that you are not clicking. (More details [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/incorporating-feedback-in-your-pull-request#applying-a-suggested-change).)

File: Makefile
Comment (brano2) [reply Makefile:74]: No, it's fine to have WIP on `develop` (as long as it doesn't break anything). I just mentioned it in the PR description and created a [ClickUp task](https://app.clickup.com/t/44fxtj) for it so we can get back to it later.

**Comment by mkristien** [comment]
Can this go? What remains to be resolved before getting an approval?

**Comment by brano2** [comment]
Sorry, it just hangs on me finding time to finish my review. (Or if you can find someone else confident enough to approve it, that would also work.) I'll be able to do it by tomorrow evening.

File: Makefile
Comment (brano2) [reply Makefile]: Wanna change it then? :D

File: Makefile
Comment (brano2) [inline Makefile]:
```suggestion
.PHONY: help
```

File: utils/build/build.mk
Comment (brano2) [inline utils/build/build.mk:7]: I have no idea what this line is/does. Could you add a short comment giving some hints/pointers? I'm not even sure where to start googling. Does DEPFLAGS stand for dependency flags? What are all the `-M<...>` options? Are they for g++?

File: utils/build/libs.mk
Comment (brano2) [reply utils/build/libs.mk]: @mkristien so should this be changed or not? Is there a reason to have it different than below?

**Review by brano2 — CHANGES_REQUESTED** [review]
I added some minor comments to the code, mostly reopening some conversations which were marked resolved without any actual resolution visible.

In addition, having played around with the makefiles, there is a couple of issues I would like addressed before this is merged:
1. As I [mentioned on Slack](https://hypedunlimited.slack.com/archives/CQ5HQJVL7/p1584196658016700?thread_ts=1582983136.013200&cid=CQ5HQJVL7), running `make RELEASE=1` produces an unhelpful error:
```utils/build/build.mk:63: *** multiple target patterns. Stop.```
If this is supposed to work without an error, can you please fix it? If not, can you document it somewhere or make it produce some more reasonable output?

2. As I also [mentioned on Slack](https://hypedunlimited.slack.com/archives/CQ5HQJVL7/p1584199148017000?thread_ts=1582983136.013200&cid=CQ5HQJVL7), there seems to be a new warning:
```Compiling src/utils/io/spi.cpp
src/utils/io/spi.cpp:34:0: warning: "_IOW" redefined
 #define _IOW(type, nr, size) 10   // random demo functionality
In file included from /usr/include/x86_64-linux-gnu/asm/ioctl.h:1:0,
                 from /usr/include/linux/ioctl.h:5,
                 from /usr/include/asm-generic/ioctls.h:5,
                 from /usr/include/x86_64-linux-gnu/asm/ioctls.h:1,
                 from /usr/include/x86_64-linux-gnu/bits/ioctls.h:23,
                 from /usr/include/x86_64-linux-gnu/sys/ioctl.h:26,
                 from src/utils/io/spi.cpp:27:
/usr/include/asm-generic/ioctl.h:85:0: note: this is the location of the previous definition
 #define _IOW(type,nr,size) _IOC(_IOC_WRITE,(type),(nr),(_IOC_TYPECHECK(size)))
```
What's up with that?

3. As [discussed in this thread on Slack](https://hypedunlimited.slack.com/archives/CQ5HQJVL7/p1582587314003100?thread_ts=1582542999.001300&cid=CQ5HQJVL7), this PR breaks `make coverage`. This seems to be because `make` and `make test` no longer put the object files in `bin/debug` and `bin/release` but instead directly into `bin`. Why this change?
One potential fix seems to be to change the paths in `test/utils/get_code_cov.sh` from `bin/debug` to `bin`. I.e. that fixed it for me but idk if it is the right way to do it. Also, the old makefiles produce `main.test.gcda` and `main.test.gcno` in `bin/debug` but the new makefiles don't seem to produce anything like that. What's happening there?
Anyway, @mkristien and @Thanakrit-Anutrakulchai, please fix `make coverage` in a way that you are both happy with.

File: utils/build/build.mk
Comment (mkristien) [reply utils/build/build.mk:7]: This is to let `make` know what needs to be recompiled. When you modify a header file, you really should recompile all object files that depend on the header file. This means not only .cpp files including the header file, but .cpp file including other headers that can eventually include the modified header file. Since the dependency tree is hard to track (for makefile), and is language (c++) specific, `DEPFLAGS` are passed to the compiler to generate those `.d` files. Those `.d` are formatted such, that they create dependency targets when included into `Makefile`. We include all of those here `Makefile:92: -include $(OBJS:.o=.d)`.

The `-M<...>` options are magic I do not entirely understand, but it works :)

To me, it seems confusing to put this explanation in the makefile. Can you suggest a way to document this? Updating the FAQ on Makefiles? It it even necessary to document?

File: utils/build/build.mk
Comment (brano2) [inline utils/build/build.mk:7]:
```suggestion
DEPFLAGS  = -MT $@ -MMD -MP -MF $(OBJS_DIR)/$*.d
# DEPFLAGS ensures that changes to headers trigger recompilation properly. More info on the wiki:
# https://github.com/Hyp-ed/hyped-2020/wiki/Makefiles#what-is-the-line-depflags----mt---mmd--mp--mf-objs_dird-in-buildmk
```

File: utils/build/build.mk
Comment (brano2) [reply utils/build/build.mk:7]: I added the [question](https://github.com/Hyp-ed/hyped-2020/wiki/Makefiles#what-is-the-line-depflags----mt---mmd--mp--mf-objs_dird-in-buildmk) to the wiki. I copied your answer above over there so feel free to modify it if you think anything could be improved.
I also suggested adding a small comment pointing to the wiki. But you might have a better formulation for that as well so feel free to change it.

**Comment by mkristien** [comment]
Points 1. 2. 3. all seem to be fixed now.

I see all comments as fixed. Please, let me know if the PR is still missing something.

File: utils/build/build.mk
Comment (mkristien) [reply utils/build/build.mk:7]: Btw, what is the policy regarding small changes like this? I think you could have committed the comments yourself directly. Are only the PR authors allowed to make changes? And reviewers are only able to suggest?

**Review by brano2 — APPROVED** [review]
Ok, looks good to me, thanks!

Normally, I'd ask for @Thanakrit-Anutrakulchai to also approve the updates relevant to QA, but since this has been open for ages, I'll be a bit of a dick, use my executive powers, and say "go ahead, merge it!" :smile:
Hope you don't mind @Thanakrit-Anutrakulchai. :sweat_smile: I think Martin has some improvements in mind for `test.mk` anyway, so if we missed anything now, there can be another (smaller) PR soon.

Also, @mkristien, for completeness (and because I'm curious), could you answer my questions in the point 3 of my previous review?

**Comment by mkristien** [comment]
Point 3. answers:
Previously, `bin/debug` was for less optimise compilations, where things like coverage can be tracked. Then, we could use `bin/release` to really squeeze the most performance out of that source code. The default was the `release` version, resulting in e.g. `pre-commit` hook compiling everything twice, once for `make` (release) and once for `make test` (debug).

The refactoring makes these changes:
- make `debug` the default
- move `bin/debug` to just `bin/`

This way, normally, we would build in `debug` and we only need to compile things once. Then, on the competition day, or if we do not need the coverage, we can turn `release` explicitly on.

**Pull request merged**
